README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
# rdecision

<!-- badges: start -->
[![CRAN status](https://www.r-pkg.org/badges/version/rdecision)](https://CRAN.R-project.org/package=rdecision) [![Codecov test coverage](https://codecov.io/gh/ajsims1704/rdecision/branch/master/graph/badge.svg)](https://codecov.io/gh/ajsims1704/rdecision?branch=master) <!-- badges: end -->

The goal of `rdecision` is to provide methods for assessing health care interventions using cohort models (decision trees). Mechanisms are provided for associating an uncertainty distribution with each source variable and for ensuring transparency of the mathematical relationships between variables. The package terminology follows Briggs *et al* "Decision Modelling for Health Economic Evaluation"<sup>1</sup>.

## Installation

You can install the released version of rdecision from [CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("rdecision")
```

## Example

Consider the fictitious and idealized decision problem of choosing between providing two forms of lifestyle advice, offered to people with vascular disease, which reduce the risk of needing an interventional procedure. The model has a time horizon of 1 year. The cost to a healthcare provider of the interventional procedure (e.g. inserting a stent) is 5000 GBP; the cost of providing the current form of lifestyle advice, an appointment with a dietician ("diet"), is 50 GBP and the cost of providing an alternative form, attendance at an exercise programme ("exercise"), is 500 GBP. If the advice programme is successful, there is no need for an interventional procedure. In a small trial of the "diet" programme, 12 out of 68 patients (17.6%) avoided having a procedure, and in a separate small trial of the "exercise" programme 18 out of 58 patients (31.0%) avoided the procedure. It is assumed that the baseline characteristics in the two trials were comparable, that the model is from the perspective of the healthcare provider and that the utility is the same for all patients.

A decision tree can be constructed to estimate the uncertainty of the cost difference between the two types of advice programme, due to the finite sample sizes of each trial. The proportions of each advice programme being successful (i.e. avoiding intervention) are represented by model variables with uncertainties which follow Beta distributions. Probabilities of the failure of the programmes are calculated using expression model variables to ensure that the total probability associated with each chance node is one.

``` r
library("rdecision")
# probabilities of programme success & failure
p.diet <- BetaModVar$new("P(diet)", "", alpha=12, beta=68-12)
p.exercise <- BetaModVar$new("P(exercise)", "", alpha=18, beta=58-18)
q.diet <- ExprModVar$new("1-P(diet)", "", rlang::quo(1-p.diet))
q.exercise <- ExprModVar$new("1-P(exercise)", "", rlang::quo(1-p.exercise))
# costs
c.diet <- 50
c.exercise <- ConstModVar$new("Cost of exercise programme", "GBP", 500)
c.stent <- 5000
```

The decision tree is constructed from nodes and edges as follows:

``` r
t.ds <- LeafNode$new("no stent")
t.df <- LeafNode$new("stent")
t.es <- LeafNode$new("no stent")
t.ef <- LeafNode$new("stent")
c.d <- ChanceNode$new("Outcome")
c.e <- ChanceNode$new("Outcome")
d <- DecisionNode$new("Programme")

e.d <- Action$new(d, c.d, cost=c.diet, label = "Diet")
e.e <- Action$new(d, c.e, cost=c.exercise, label = "Exercise")
e.ds <- Reaction$new(c.d, t.ds, p=p.diet, cost = 0, label = "success")
e.df <- Reaction$new(c.d, t.df, p=q.diet, cost=c.stent, label="failure")
e.es <- Reaction$new(c.e, t.es, p=p.exercise, cost=0, label="success")
e.ef <- Reaction$new(c.e, t.ef, p=q.exercise, cost=c.stent, label="failure")

DT <- DecisionTree$new(
  V = list(d, c.d, c.e, t.ds, t.df, t.es, t.ef),
  E = list(e.d, e.e, e.ds, e.df, e.es, e.ef)
)
```

<img src="man/figures/README-treedraw-1.png" width="75%" style="display: block; margin: auto;" />

The expected per-patient net cost of each option is obtained by evaluating the tree with expected values of all variables using `DT$evaluate()` and threshold values with `DT$threshold()`. Examination of the results of evaluation shows that the expected per-patient net cost of the diet advice programme is 4167.65 GBP and the per-patient net cost of the exercise programme is 3948.28 GBP, a point estimate saving of 219.37 GBP per patient if the exercise advice programme is adopted. By univariate threshold analysis, the exercise program will be cost saving when its cost of delivery is less than 719.73 GBP or when its success rate is greater than 26.6%.

The confidence interval of the cost saving is estimated by repeated evaluation of the tree, each time sampling from the uncertainty distribution of the two probabilities using, for example, `DT$evaluate(setvars="random", N=1000)` and inspecting the resulting data frame. From 1000 runs, the 95% confidence interval of the per patient cost saving is -496.79 GBP to 950.36 GBP, with 73.1% being cost saving, and it can be concluded that more evidence is required to be confident that the exercise programme is cost saving.

# Acknowledgements

In addition to using base R<sup>2</sup>, `redecision` relies heavily on the `R6` implementation of classes<sup>3</sup> and the `rlang` package for error handling and non-standard evaluation used in expression model variables<sup>4</sup>. Building the package vignettes and documentation relies on the `testthat` package<sup>5</sup>, the `devtools` package<sup>6</sup> and `rmarkdown`<sup>7</sup>.

Underpinning graph theory is based on terminology, definitions and algorithms from Gross *et al*<sup>10</sup>, the Wikipedia glossary<sup>11</sup> and links therein. Topological sorting of graphs is based on Kahn's algorithm<sup>12</sup>. Some of the terminology for decision trees was based on the work of Kaminski *et al*<sup>13</sup> and an efficient tree drawing algorithm was based on the work of Walker<sup>14</sup>. Terminology for decision trees in health economic evaluation was based on the book by Briggs *et al*<sup>1</sup> and the output format and terminology follows ISPOR recommendations<sup>15</sup>.

Citations for examples used in vignettes are given in applicable vignette files.

# References

1 Briggs A, Claxton K, Sculpher M. *Decision modelling for health economic evaluation*. Oxford, UK: Oxford University Press; 2006.

2 R Core Team. *R: A language and environment for statistical computing*. Vienna, Austria: R Foundation for Statistical Computing; 2020.

3 Chang W. *R6: Encapsulated classes with reference semantics*. 2020.

4 Henry L, Wickham H. *Rlang: Functions for base types and core r and ’tidyverse’ features*. 2020.

5 Wickham H. Testthat: Get started with testing. *The R Journal* 2011;**3**:5–10.

6 Wickham H, Hester J, Chang W. *Devtools: Tools to make developing r packages easier*. 2020.

7 Xie Y, Allaire J, Grolemund G. *R markdown: The definitive guide*. Boca Raton, Florida: Chapman and Hall/CRC; 2018.

8 Allaire J, Xie Y, McPherson J, Luraschi J, Ushey K, Atkins A *et al.* *Rmarkdown: Dynamic documents for r*. 2020.

9 Xie Y, Dervieux C, Riederer E. *R markdown cookbook*. Boca Raton, Florida: Chapman and Hall/CRC; 2020.

10 Gross JL, Yellen J, Zhang P. *Handbook of Graph Theory*. 2nd ed. Chapman and Hall/CRC.; 2013.

11 Wikipedia. Glossary of graph theory. *Wikipedia* 2021.

12 Kahn AB. Topological sorting of large networks. *Communications of the ACM* 1962;**5**:558–62. <https://doi.org/10.1145/368996.369025>.

13 Kamiński B, Jakubczyk M, Szufel P. A framework for sensitivity analysis of decision trees. *Central European Journal of Operational Research* 2018;**26**:135–59. <https://doi.org/10.1007/s10100-017-0479-6>.

14 Walker JQ. *A node-positioning algorithm for general trees*. Chapel Hill: University of North Carolina; 1989.

15 Briggs AH, Weinstein MC, Fenwick EA, Karnon J, Sculpher MJ, Paltiel AD. Model Parameter Estimation and Uncertainty: A Report of the ISPOR-SMDM Modeling Good Research Practices Task Force-6. *Value in Health* 2012;**15**:835–42. <https://doi.org/10.1016/j.jval.2012.04.014>.


inst/NEWS.md

# rdecision 1.0.4

* Added option `value` to method `set` in class `ModVar`. This allows
  variables to be set to an explicit value; used in threshold finding.
* Added `threshold` function to `DecisionTree` to calculate the value of a 
  model variable at which the cost difference becomes zero or ICER crosses
  a threshold.
* Added option `run` to `by` argument of `DecisionTree$evaluate()`. Avoids
  application having to `reshape` output before reporting PSA results.
* Fixed bug in method `DecisionTree$tornado` which caused bars to be clipped 
  under some circumstances.
* Minor revisions to the Tegaderm vignette.

# rdecision 1.0.3

* Package tests that involve sampling randomly from a distribution and
  comparing the results with parameters of an expected distribution have been
  excluded when running CRAN tests. Otherwise the central limit theorem or 
  empirical distributions are used to find 99.9% confidence limits on sample
  mean and SD.
* Added common test helpers and bespoke expectations to `testthat/setup.R`.
* Changed vignette titles to reflect what kind of problem they illustrate,
  rather than the problems themselves, to make it clearer on the CRAN page.
* Added method `as_DOT` to `Graph` and `Digraph` for export to graphviz DOT
  file format to aid visualization of graphs.

# rdecision 1.0.2

* Added tests to give 100% coverage and replaced `tolerance` in
  `expect_equal` with `abs` in expect_true for approximate equality tests.
* Further description for documentation.
* Converted vignettes to HTML.
* Added `WORDLIST` file and sundry administrative changes for clean package
  build.
* Added `README` file, with an example and acknowledgements.

# rdecision 1.0.1

* Added `draw()` method to `DecisionTree`.
* Added `tornado()` method to `DecisionTree` for univariate sensitivity 
  analysis.
* Optimized probabilistic sensitivity analysis loop in `DecisionTree` 
  (1000 evaluations of a typical HTA tree takes < 5s on a typical PC).

# rdecision 1.0.0

* First full release of the package.
* Added graph theory classes. Decision trees and Markov models are
  forms of graph.
* Renamed `ModelVariable` as `ModVar` for compactness, and renamed its derived
  classes similarly.
* Added test harnesses for more classes.
* Collected vignette citations to file references.bib and changed 
  to *BMJ* csl style.
* Added extra graph theory and decision tree vignettes.  
  
# rdecision 0.1.7

* Removed the label argument from `ModelVariable`. 
* Improved auto-detection of variable label in `ModelVariable`.
* Added NEWS.md and `CITATION` file to inst folder in CRAN preparation.
* Added `tests/testthat` folder with tests for `ModelVariable`.
* Added scripts to call devtools::check/build on Windows/Mac.
* Fixed notes issued by R CMD check.

# rdecision 0.1.6

* Introduced the `ModelVariable` class as the new base class from which to 
  construct the variables in an economic model. The class includes
  methods to support parametrization of uncertainty in the model variable.
* Introduced sub-classes of `ModelVariable` to model particular forms
  of uncertainty. These are `ConstModelVariable`, `NormalModelVariable`,
  `GammaModelVariable`, `BetaModelVariable`, `LogNormalModelVariable`. They
  do as expected from their names. Some support alternative forms
  of parametrization.
* Introduced `ExpressionModelVariable`. A sub-class of `ModelVariable`, objects
  of this class are defined with an expression involving other model 
  variables. The concept permits variables to be combined in any mathematical
  expression that R itself will support. Because `ExpressionModelVariable`s are
  themselves `ModelVariables`, they can can appear in an expression that 
  is used to define another model variable.
* Introduced tabulation functions to list the properties of a model variable
  and its operands.
* Revised `Node` and its sub classes to accept `ModelVariables` as arguments
  to costs, utilities and probabilities, thus embedding probabilistic
  sensitivity analysis into decision tree models.
* Added the Tegaderm vignette. This is a published example of a decision tree
  model with PSA and is partial validation of the `ModelVariable` approach
  to PSA.
* Updated the Sumatriptan vignette, after subsuming some of its pathway
  traversal code into `Node` classes.
* Removed `node.apply` and `path.apply` functions, and subsumed them into 
  `Node`.
* Removed functions intended for use with `node.apply` and `path.apply`, and
  subsumed them into `Node`.
* Provided `Node` objects with a Document Object Model (DOM) interface, as
  far as practicable.
  
# rdecision 0.1.3

* Moved citations in vignettes from external file `references.bib` to directly
  embed them in the YAML headers. To do: explore whether references can be
  saved in preferred bib format.
* Replaced call to `nullfile()`, for suppressed output, in function `des` with
  detection of OS to support older R versions (`nullfile` was introduced to
  base R at 3.6.0).

# rdecision 0.1.2

* For the Markov solver:
    * Function is now called `des`
    * It returns a list of summary matrices (the same ones written to csv files) 
      instead of a single number.
    * Output can be suppressed by setting stub=NA.
    * Some minor bugs fixed.
    
# rdecision 0.1.1

* First local release of rdecision as a package.
* Added classes for solving decision trees (`Node`, `LeafNode`, `ChanceNode`,
  `DecisionNode`) and pathway detection and traversal functions.
* Incorporated our discrete event solver, originally written in Matlab
  for the WatchBP model, then translated as a stand-alone R script, into
  the package.
* Added vignettes for Sumatriptan model from Briggs (Box 2.3) and
  from Sonnenberg and Beck's original 3-state example.


R/ExprModVar.R
#' @title \verb{ExprModVar} class
#' 
#' @description 
#' An R6 class for a model variable constructed from an expression
#' involving other model variables.
#' 
#' @details A class to support expressions involving objects
#' of base class \code{ModVar}, which itself behaves like a 
#' model variable. For example, if \code{A} and
#' \code{B} are variables with base class \code{ModVar}
#' and \code{c} is a variable of type \code{numeric}, then
#' it is not possible to write, for example, 
#' \code{x <- 42*A/B + c}, because R cannot manipulate class
#' variables using the same operators as regular variables. 
#' But such forms of expression may be desirable in constructing a
#' model and this class provides a mechanism for doing so.
#' 
#' @references{
#'   Briggs A, Claxton K, Sculpher M. Decision modelling for health
#'   economic evaluation. Oxford, UK: Oxford University Press; 2006. 
#' }
#' 
#' @note For many expressions involving model variables there will 
#' be no closed form expressions for the mean, standard deviation and
#' the quantiles. Therefore they are obtained by simulation, via
#' functions \code{mu_hat}, \code{sigma_hat} and \code{q_hat}.
#' 
#' @note For consistency with \code{ModVar}s which are not expressions, the
#' function \code{mean} returns the value of the expression when all
#' its operands take their mean values. This will, in general, not
#' be the mean of the expression distribution (which can be obtained
#' via \code{mu_hat}), but is the value normally used in the base
#' case of a model as the point estimate. As Briggs \emph{et al} note 
#' (section 4.1.1) "in all but the most non-linear models, the 
#' difference between the expectation over the output of a 
#' probabilistic model and that model evaluated at the mean values
#' of the input parameters, is likely to be modest."
#' 
#' @note Functions \code{SD}, \code{mode} and \code{quantile} return \code{NA}
#' because they do not necessarily have a closed form. The standard
#' deviation can be estimated by calling \code{sigma_hat} and the
#' quantiles by \code{q_hat}. Because a unimodal distribution is not
#' guaranteed, there is no estimator provided for the mode.
#' 
#' @note Method \code{distribution} returns the string representation 
#' of the expression used to create the model variable.
#' 
#' @author Andrew J. Sims \email{andrew.sims@@newcastle.ac.uk}
#' @docType class
#' @export
#' 
ExprModVar <- R6::R6Class(
  classname = "ExprModVar",
  lock_class = TRUE,
  inherit =  ModVar,
  private = list(
    # fields
    expr = NULL,
    env = NULL,
    q.mean = NULL,
    q.r1 = NULL,
    q.get = NULL
  ),
  public = list(
    
    #' @description 
    #' Create a \code{ModVar} formed from an expression involving other
    #' model variables.
    #' @param description Name for the model variable expression. In 
    #' a complex model it may help to tabulate how model variables are
    #' combined into costs, probabilities and rates.
    #' @param units Units in which the variable is expressed.
    #' @param quo A \verb{quosure} (see package \pkg{rlang}), which contains an 
    #' expression and its environment. The usage is \code{quo(x+y)} or 
    #' \code{rlang::quo(x+y)}.
    #' @return An object of type \code{ExprModVar}
    initialize = function(description, units, quo) {
      # initialize the base class
      super$initialize(description, units)
      # check,split and save the quosure
      if (!rlang::is_quosure(quo)) {
        rlang::abort("Argument quo must be a quosure", class="quo_not_quosure")
      }
      private$expr <- rlang::quo_get_expr(quo)
      private$env <- rlang::quo_get_env(quo)
      # create pre-packaged evaluable quosures to optimize calculations
      private$q.mean <- self$add_method("mean()")
      private$q.r1 <- self$add_method("r(1)")
      private$q.get <- self$add_method("get()")
      # return
      return(invisible(self))
    },

    #' @description 
    #' Create a new \verb{quosure} from that supplied in \code{new()} but with
    #' each \code{ModVar}
    #' operand appended with \code{$x} where \code{x} is the argument to this
    #' function.
    #' @param method A character string with the method, e.g. \code{"mean()"}.
    #' @return A \dfn{quosure} whose expression is each \code{ModVar} \code{v}
    #' in the  
    #' expression replaced with \code{v$method} and the same environment as
    #' specified in the quosure supplied in new().
    #' @details This method is mostly intended for internal use within the
    #' class and will not generally be needed for normal use of
    #' \code{ExprModVar} objects. The returned expression is \emph{not} 
    #' syntactically checked or evaluated before it is returned.
    add_method = function(method="mean()") {
      # check argument
      if (!is.character(method)) {
        rlang::abort(
          "Argument 'method' must be a character string",
          class = "method_not_character"
        )
      }
      # substitute model variable names with call to method
      mv <- list()
      for (v in all.vars(private$expr)) {
        # build replacement string if necessary
        if (inherits(eval(str2lang(v), envir=private$env), what='ModVar')) {
          rep <- gsub(v, paste(v, method, sep='$'), v)
        }
        else {
          rep <- v
        }
        # create symbol from the modified string
        mv[[v]] <- str2lang(rep)
      }
      emod <- eval(substitute(substitute(e, env=mv), env=list(e=private$expr)))
      # create new quosure
      q <- rlang::new_quosure(expr=emod, env=private$env)
      # return the quosure
      return(q)
    },
    
    #' @description 
    #' Tests whether the model variable is probabilistic, i.e. a random
    #' variable that follows a distribution, or an expression involving
    #' random variables, at least one of which follows a distribution. 
    #' @return \code{TRUE} if probabilistic
    is_probabilistic = function() {
      # test each operand for whether it is probabilistic
      lv <- sapply(base::all.vars(private$expr), FUN=function(v) {
        rv <- as.logical(NA)
        vv <- rlang::eval_bare(expr=rlang::sym(v), env=private$env) 
        if (inherits(vv, what="ModVar")) {
          rv <- vv$is_probabilistic()
        } else if (inherits(vv, what="numeric")) {
          rv <- FALSE
        }
        return(rv)
      })
      return(any(lv))
    },

    #' @description
    #' Return a list of operands that are themselves \code{ModVar}s given
    #' in the expression.
    #' @return A list of model variables.
    operands = function() {
      # filter the expression variables that are ModVars
      mvlist <- list()
      for (v in all.vars(private$expr)) {
        vv <- rlang::eval_bare(expr=rlang::sym(v), env=private$env) 
        if (inherits(vv, what="ModVar")) {
          # add the variable to the list
          mvlist <- c(mvlist, vv)
          # and add its operands, if any
          if (inherits(vv, what="ExprModVar")) {
            for (o in vv$operands()) {
              mvlist <- c(mvlist, o)
            }
          }
        }
      }
      return(unique(mvlist))
    },
    
    #' @description 
    #' Accessor function for the name of the expression model variable.
    #' @return Expression as a character string with all control characters
    #' having been removed.
    distribution = function() {
      estr <- rlang::as_label(
        rlang::new_quosure(expr = private$expr, env = private$env)
      )
      estr <- gsub("[[:cntrl:]]", "", estr)
      return(estr)
    },

    #' @description 
    #' Draw a random sample from the model variable. 
    #' @param n Number of samples to draw.
    #' @return A sample drawn at random.
    r = function(n=1) {
      rv <- NA
      # if n=1 use the pre-packed expression
      if (n==1) {
        rv <- rlang::eval_tidy(private$q.r1)
      } else {
        # build expression with each variable replaced with $r(n)
        method <- paste0("r(", n, ")")
        q.r <- self$add_method(method)
        # evaluate the expression
        rv <- rlang::eval_tidy(q.r)
      }
      # return result
      return(rv)
    },

    #' @description 
    #' Return the value of the expression when its operands take their
    #' mean value (i.e. value returned by call to \code{mean} or their
    #' value, if numeric). See notes on this class for further explanation.
    #' @return Mean value as a numeric value.
    mean = function() {
      # evaluate the pre-packaged expression
      rv <- rlang::eval_tidy(private$q.mean)
      # return it
      return(rv)
    },
    
    #' @description 
    #' Return the mode of the variable. By default returns \code{NA}, which 
    #' will be the case for most \code{ExprModVar} variables, because an 
    #' arbitrary expression is not guaranteed to be unimodal.
    #' @return Mode as a numeric value.
    mode = function() {
      return(as.numeric(NA))
    },

    #' @description 
    #' Return the standard deviation of the distribution as \code{NA} because 
    #' the variance is not available as a closed form for all functions of
    #' distributions. 
    #' @return Standard deviation as a numeric value
    SD = function() {
      return(as.numeric(NA))
    },
    
    #' @description 
    #' Find quantiles of the uncertainty distribution.  Not available
    #' as a closed form, and returned as \code{NA}.
    #' @param probs Numeric vector of probabilities, each in range [0,1].
    #' @return Vector of numeric values of the same length as \code{probs}.
    quantile = function(probs) {
      # test argument
      sapply(probs, FUN=function(x) {
        if (is.na(x)) {
          rlang::abort("All elements of 'probs' must be defined",
                       class="probs_not_defined")
        }
        if (!is.numeric(x)) {
          rlang::abort("Argument 'probs' must be a numeric vector",
                       class="probs_not_numeric")
        }
        if (x<0 || x>1) {
          rlang::abort("Elements of 'probs' must be in range[0,1]",
                       class="probs_out_of_range")
        }
      })
      return(rep(as.numeric(NA),length(probs)))
    },

    #' @description 
    #' Return the estimated expected value of the expression variable. This is
    #' computed by numerical simulation because there is, in general, no closed
    #' form expressions for the mean of a function of distributions.
    #' @param nest Sample size to be used to estimate the mean. Values less than
    #' 1000 (default) are unlikely to return meaningful estimates and will 
    #' be rejected.
    #' @return Expected value as a numeric value.
    mu_hat = function(nest=1000) {
      # check and process the sample estimation size
      if (!is.numeric(nest)) {
        rlang::abort("Argument nest must be numeric", class="nest_not_numeric")
      }
      if (nest < 1000) {
        rlang::abort(
          "Argument nest must not be less than 1000", 
          class="nest_too_small"
        )
      }
      # sample values from the variable
      S <- self$r(n=nest)
      return(mean(S))
    },
    
    #' @description 
    #' Return the estimated standard deviation of the distribution. This is
    #' computed by numerical simulation because there is, in general, no closed
    #' form expressions for the SD of a function of distributions.
    #' @param nest Sample size to be used to estimate the SD. Values less than
    #' 1000 (default) are unlikely to return meaningful estimates and will 
    #' be rejected.
    #' @return Standard deviation as a numeric value.
    sigma_hat = function(nest=1000) {
      # check and process the sample estimation size
      if (!is.numeric(nest)) {
        rlang::abort("Argument nest must be numeric", class="nest_not_numeric")
      }
      if (nest < 1000) {
        rlang::abort(
          "Argument nest must not be less than 1000", 
          class="nest_too_small"
        )
      }
      # sample values from the variable
      S <- self$r(nest)
      return(sd(S))
    },

    #' @description
    #' Return the estimated quantiles by sampling the variable. This is
    #' computed by numerical simulation because there is, in general, no closed
    #' form expressions for the quantiles of a function of distributions.
    #' @param probs Vector of probabilities, in range [0,1].    
    #' @param nest Sample size to be used to estimate the SD. Values less than
    #' 1000 (default) are unlikely to return meaningful estimates and will 
    #' be rejected.
    #' @return Vector of quantiles.
    q_hat = function(probs, nest=1000) {
      # test probs
      sapply(probs, FUN=function(x) {
        if (is.na(x)) {
          rlang::abort("All elements of 'probs' must be defined",
                       class="probs_not_defined")
        }
        if (!is.numeric(x)) {
          rlang::abort("Argument 'probs' must be a numeric vector",
                       class="probs_not_numeric")
        }
        if (x<0 || x>1) {
          rlang::abort("Elements of 'probs' must be in range[0,1]",
                       class="probs_out_of_range")
        }
      })
      # check and process the sample estimation size
      if (!is.numeric(nest)) {
        rlang::abort("Argument nest must be numeric", class="nest_not_numeric")
      }
      if (nest < 1000) {
        rlang::abort(
          "Argument nest must not be less than 1000", 
          class="nest_too_small"
        )
      }
      # sample the distribution
      S <- self$r(nest)
      # return the quantiles of the sample
      q <- quantile(S, probs)
      return(q)
    },
    
    #' @description
    #' Sets the value of the \code{ExprModVar} that will be returned by 
    #' subsequent calls to \code{get()} until \code{set()} is called again. 
    #' Because an \code{ExprModVar} can be considered as a dependent variable,
    #' the idea of \code{set}ting a value is meaningless, and calls to this
    #' method have no effect. To affect the
    #' value returned by the next call to \code{get}, call \code{set} for each
    #' of the operands of this expression.
    #' @param what Character string; for compatibility with non-expression
    #' \code{ModVar}s only; not used.
    #' @param val Numerical value when what="value"; not used for expression
    #' model variables.
    #' @return Updated \code{ExprModVar}.
    set = function(what="random", val=NULL) {
      # check argument
      if (!is.character(what)) {
        rlang::abort(
          "Argument 'what' must be a character", 
          class = "what_not_character"
        )
      }
      opts <- c("random","expected","q2.5","q50","q97.5","current", "value")
      if (!(what %in% opts)) {
        rlang::abort(
          paste("'what' must be one of", paste(opts, collpse= "|")), 
          class ="what_not_supported"
        )
      }
      return(invisible(self))
    },
    
    #' @description
    #' Gets the value of the \code{ExprModVar} that was set by the most recent
    #' call to \code{set()} to each operand of the expression.
    #' @return Value determined by last \code{set()}.
    get = function() {
      # evaluate the pre-packaged expression
      rv <- rlang::eval_tidy(private$q.get)
      # return it
      return(rv)
    }

  )
)
